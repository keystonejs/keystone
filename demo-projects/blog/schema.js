require('dotenv').config();
const {
  File,
  Text,
  Slug,
  Relationship,
  Select,
  Password,
  Checkbox,
  CalendarDay,
  DateTime,
  OEmbed,
} = require('@keystonejs/fields');
const { Wysiwyg } = require('@keystonejs/fields-wysiwyg-tinymce');
const { LocalFileAdapter } = require('@keystonejs/file-adapters');
const getYear = require('date-fns/get_year');

const { staticRoute, staticPath, distDir } = require('./config');
const dev = process.env.NODE_ENV !== 'production';

let iframelyAdapter;

if (process.env.IFRAMELY_API_KEY) {
  const { IframelyOEmbedAdapter } = require('@keystonejs/oembed-adapters');
  iframelyAdapter = new IframelyOEmbedAdapter({
    apiKey: process.env.IFRAMELY_API_KEY,
  });
}

const fileAdapter = new LocalFileAdapter({
  src: `${dev ? '' : `${distDir}/`}${staticPath}/uploads`,
  path: `${staticRoute}/uploads`,
});

const avatarFileAdapter = new LocalFileAdapter({
  src: `${staticPath}/avatars`,
  path: `${staticRoute}/avatars`,
});

exports.User = {
  fields: {
    name: { type: Text },
    email: { type: Text, isUnique: true },
    dob: {
      type: CalendarDay,
      format: 'Do MMMM YYYY',
      yearRangeFrom: 1901,
      yearRangeTo: getYear(new Date()),
    },
    ...(process.env.IFRAMELY_API_KEY
      ? {
          portfolio: { type: OEmbed, adapter: iframelyAdapter },
        }
      : {}),
    password: { type: Password },
    isAdmin: { type: Checkbox },
    avatar: { type: File, adapter: avatarFileAdapter },
  },
  labelResolver: item => `${item.name} <${item.email}>`,
};

exports.Post = {
  fields: {
    title: { type: Text },
    slug: { type: Slug, from: 'title' },
    author: {
      type: Relationship,
      ref: 'User',
    },
    categories: {
      type: Relationship,
      ref: 'PostCategory',
      many: true,
    },
    status: {
      type: Select,
      defaultValue: 'draft',
      options: [
        { label: 'Draft', value: 'draft' },
        { label: 'Published', value: 'published' },
      ],
    },
    body: { type: Wysiwyg },
    posted: { type: DateTime, format: 'DD/MM/YYYY' },
    image: { type: File, adapter: fileAdapter },
  },
  adminConfig: {
    defaultPageSize: 20,
    defaultColumns: 'title, status',
    defaultSort: 'title',
  },
  labelResolver: item => item.title,
};

exports.PostCategory = {
  fields: {
    name: { type: Text },
    slug: { type: Slug, from: 'name' },
  },
};

exports.Comment = {
  fields: {
    body: { type: Text, isMultiline: true },
    originalPost: {
      type: Relationship,
      ref: 'Post',
    },
    author: {
      type: Relationship,
      ref: 'User',
    },
    posted: { type: DateTime },
  },
  labelResolver: item => item.body,
};
