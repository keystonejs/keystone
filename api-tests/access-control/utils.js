const { setupServer } = require('@keystonejs/test-utils');
const { Text } = require('@keystonejs/fields');
const { objMerge } = require('@keystonejs/utils');

const FAKE_ID = { mongoose: '5b3eabd9e9f2e3e4866742ea', knex: 137 };
const FAKE_ID_2 = { mongoose: '5b3eabd9e9f2e3e4866742eb', knex: 138 };

const yesNo = truthy => (truthy ? 'Yes' : 'No');

const getPrefix = access => {
  // prettier-ignore
  let prefix = `${yesNo(access.create)}Create${yesNo(access.read)}Read${yesNo(access.update)}Update${yesNo(access.auth)}Auth`;
  if (Object.prototype.hasOwnProperty.call(access, 'delete')) {
    prefix = `${prefix}${yesNo(access.delete)}Delete`;
  }
  return prefix;
};

const getStaticListName = access => `${getPrefix(access)}StaticList`;
const getImperativeListName = access => `${getPrefix(access)}ImperativeList`;
const getDeclarativeListName = access => `${getPrefix(access)}DeclarativeList`;

/* Generated with:
  const result = [];
  const options = ['create', 'read', 'update', 'delete', 'auth'];
  // All possible combinations are contained in the set 0..2^n-1
  for(let flags = 0; flags < Math.pow(2, options.length); flags++) {
    // Generate an object of true/false values for the particular combination
    result.push(options.reduce((memo, option, index) => ({
      ...memo,
      // Use a bit mask to see if that bit is set
      [option]: !!(flags & (1 << index)),
    }), {}));
  }
  */
const listAccessVariations = [
  { create: false, read: false, update: false, delete: false, auth: true },
  { create: true, read: false, update: false, delete: false, auth: true },
  { create: false, read: true, update: false, delete: false, auth: true },
  { create: true, read: true, update: false, delete: false, auth: true },
  { create: false, read: false, update: true, delete: false, auth: true },
  { create: true, read: false, update: true, delete: false, auth: true },
  { create: false, read: true, update: true, delete: false, auth: true },
  { create: true, read: true, update: true, delete: false, auth: true },
  { create: false, read: false, update: false, delete: true, auth: true },
  { create: true, read: false, update: false, delete: true, auth: true },
  { create: false, read: true, update: false, delete: true, auth: true },
  { create: true, read: true, update: false, delete: true, auth: true },
  { create: false, read: false, update: true, delete: true, auth: true },
  { create: true, read: false, update: true, delete: true, auth: true },
  { create: false, read: true, update: true, delete: true, auth: true },
  { create: true, read: true, update: true, delete: true, auth: true },
  { create: false, read: false, update: false, delete: false, auth: false },
  { create: true, read: false, update: false, delete: false, auth: false },
  { create: false, read: true, update: false, delete: false, auth: false },
  { create: true, read: true, update: false, delete: false, auth: false },
  { create: false, read: false, update: true, delete: false, auth: false },
  { create: true, read: false, update: true, delete: false, auth: false },
  { create: false, read: true, update: true, delete: false, auth: false },
  { create: true, read: true, update: true, delete: false, auth: false },
  { create: false, read: false, update: false, delete: true, auth: false },
  { create: true, read: false, update: false, delete: true, auth: false },
  { create: false, read: true, update: false, delete: true, auth: false },
  { create: true, read: true, update: false, delete: true, auth: false },
  { create: false, read: false, update: true, delete: true, auth: false },
  { create: true, read: false, update: true, delete: true, auth: false },
  { create: false, read: true, update: true, delete: true, auth: false },
  { create: true, read: true, update: true, delete: true, auth: false },
];

const fieldMatrix = [
  { create: false, read: false, update: false },
  { create: true, read: false, update: false },
  { create: false, read: true, update: false },
  { create: true, read: true, update: false },
  { create: false, read: false, update: true },
  { create: true, read: false, update: true },
  { create: false, read: true, update: true },
  { create: true, read: true, update: true },
];
const getFieldName = access => getPrefix(access);

const nameFn = { imperative: getImperativeListName, declarative: getDeclarativeListName };

const createFieldStatic = fieldAccess => ({
  [getFieldName(fieldAccess)]: {
    type: Text,
    access: fieldAccess,
  },
});
const createFieldImperative = fieldAccess => ({
  [getFieldName(fieldAccess)]: {
    type: Text,
    access: {
      create: () => fieldAccess.create,
      read: () => fieldAccess.read,
      update: () => fieldAccess.update,
    },
  },
});
function setupKeystone(adapterName) {
  return setupServer({
    adapterName,
    createLists: keystone => {
      keystone.createList('User', { fields: { name: { type: Text } } });

      listAccessVariations.forEach(access => {
        keystone.createList(getStaticListName(access), {
          fields: {
            name: { type: Text },
            ...objMerge(fieldMatrix.map(variation => createFieldStatic(variation))),
          },
          access,
        });
        keystone.createList(getImperativeListName(access), {
          fields: {
            name: { type: Text },
            ...objMerge(fieldMatrix.map(variation => createFieldImperative(variation))),
          },
          access: {
            create: () => access.create,
            read: () => access.read,
            update: () => access.update,
            delete: () => access.delete,
            auth: () => access.auth,
          },
        });
        keystone.createList(getDeclarativeListName(access), {
          fields: { name: { type: Text } },
          access: {
            create: access.create,
            // arbitrarily restrict the data to a single item (see data.js)
            read: () => access.read && { name_starts_with: 'Hello' },
            update: () => access.update && { name_starts_with: 'Hello' },
            delete: () => access.delete && { name_starts_with: 'Hello' },
            auth: access.auth,
          },
        });
      });
    },
  });
}
module.exports = {
  FAKE_ID,
  FAKE_ID_2,
  getStaticListName,
  getImperativeListName,
  getDeclarativeListName,
  listAccessVariations,
  fieldMatrix,
  nameFn,
  setupKeystone,
  getFieldName,
};
