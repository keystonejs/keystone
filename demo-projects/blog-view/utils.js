const terminalLink = require('terminal-link');
const ciInfo = require('ci-info');
const chalk = require('chalk');

const ttyLink = (text, path, port) => {
  if (ciInfo.isCI) {
    return;
  }
  const url = `http://localhost:${port}${path}`;
  const link = terminalLink(url, url, { fallback: () => url });
  console.log(`ðŸ”— ${chalk.green(text)}\t${link}`);
};

function extractAppMeta(apps) {
  let adminPath;
  let graphiqlPath;
  let apiPath;

  apps.forEach(app => {
    switch (app.constructor.name) {
      case 'AdminUIApp': {
        adminPath = app.adminPath;
        break;
      }
      case 'GraphQLApp': {
        apiPath = app._apiPath;
        graphiqlPath = app._graphiqlPath;
        break;
      }
    }
  });

  return {
    adminPath,
    graphiqlPath,
    apiPath,
  };
}

function logAdminRoutes(apps, port) {
  const { adminPath, graphiqlPath, apiPath } = extractAppMeta(apps);
  /* eslint-disable no-unused-expressions */
  adminPath && ttyLink('Keystone Admin UI:', adminPath, port);
  graphiqlPath && ttyLink('GraphQL Playground:', graphiqlPath, port);
  apiPath && ttyLink('GraphQL API:\t', apiPath, port);
  /* eslint-enable no-unused-expressions */
}

module.exports = {
  logAdminRoutes,
};
