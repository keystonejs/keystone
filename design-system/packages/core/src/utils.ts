import {
  ComponentPropsWithoutRef,
  ElementType,
  forwardRef,
  ReactElement,
  ReactNode,
  Ref,
  useEffect,
  useLayoutEffect,
  useState,
} from 'react';
import { createPortal } from 'react-dom';

/*
  @johannes' one weird trick for fixing TypeScript autocomplete
*/
export function identityType<T>() {
  function inner<U extends T>(u: U): U {
    return u;
  }
  return inner;
}

/*
  Logs a warning to the console when the condition is true, only in dev
*/
export const devWarning = (condition: boolean, message: string) => {
  if (process.env.NODE_ENV !== 'production') {
    if (condition) {
      console.error(message);
    }
  }
};

/*
  forwardRefWithAs lets us forward refs while keeping the correct component type,
  which can be specified by the `as` prop.
*/

type ElementTagNameMap = HTMLElementTagNameMap &
  Pick<SVGElementTagNameMap, Exclude<keyof SVGElementTagNameMap, keyof HTMLElementTagNameMap>>;

type AsProp<Comp extends ElementType, Props> = {
  as?: Comp;
  ref?: Ref<
    Comp extends keyof ElementTagNameMap
      ? ElementTagNameMap[Comp]
      : Comp extends new (...args: any) => any
      ? InstanceType<Comp>
      : undefined
  >;
} & Omit<ComponentPropsWithoutRef<Comp>, 'as' | keyof Props>;

type CompWithAsProp<Props, DefaultElementType extends ElementType> = <
  Comp extends ElementType = DefaultElementType
>(
  props: AsProp<Comp, Props> & Props
) => ReactElement;

export const forwardRefWithAs = <DefaultElementType extends ElementType, BaseProps>(
  render: (
    props: BaseProps & { as?: ElementType },
    ref: React.Ref<any>
  ) => Exclude<ReactNode, undefined>
): CompWithAsProp<BaseProps, DefaultElementType> => {
  // @ts-ignore
  return forwardRef(render);
};

/*
  A helper for making valid IDs from a set of inputs
*/
export function makeId(...args: (string | number | null | undefined)[]) {
  return args.filter(val => val != null).join('--');
}

/*
  A helper for handling string OR array values e.g.

  <Component size="small" />
  VS
  <Component size={['small', 'large']} />
*/
export const mapResponsiveProp = <
  Map extends Record<string, string | number>,
  Keys extends keyof Map
>(
  value: Keys | readonly (Keys | null)[],
  valueMap: Map
) => {
  if (Array.isArray(value)) {
    return value.map(k => (k == null ? null : valueMap[k]));
  }
  // @ts-ignore
  return valueMap[value];
};

/**
 * Utils below are ported with thanks from @reach-ui
 * Copyright (c) 2018-present, React Training LLC
 */

// Autogenerate IDs to facilitate WAI-ARIA and server rendering. For reasoning, see
// https://github.com/reach/reach-ui/blob/develop/packages/auto-id/src/index.tsx
let serverHandoffComplete = false;
let id = 0;
const genId = () => ++id;

export const useId = (idFromProps?: string | null) => {
  const initialId = idFromProps || (serverHandoffComplete ? genId() : null);

  const [id, setId] = useState(initialId);

  useSafeLayoutEffect(() => {
    if (id === null) {
      setId(genId());
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (serverHandoffComplete === false) {
      serverHandoffComplete = true;
    }
  }, []);
  return id != null ? String(id) : undefined;
};

// Works around useLayoutEffect throwing a warning when used in SSR
export const useSafeLayoutEffect = typeof window === 'undefined' ? () => {} : useLayoutEffect;

type Props = {
  children: ReactElement;
};

export const Portal = ({ children }: Props) => {
  if (typeof document === 'undefined') {
    return null;
  }

  return createPortal(children, document.body);
};
