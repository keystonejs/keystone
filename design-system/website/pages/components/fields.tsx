/* @jsx jsx */

import { ReactNode, useState } from 'react';
import { jsx } from '@emotion/core';
import { useTheme } from '@keystone-ui/core';
import { Checkbox, Radio, TextArea, TextInput, Switch } from '@keystone-ui/fields';
import { SegmentedControl } from '@keystone-ui/segmented-control';

import { Page } from '../../components/Page';

const FieldWrapper = ({ children }: { children: ReactNode }) => {
  const { spacing } = useTheme();
  return (
    <div
      css={{
        marginTop: spacing.medium,
        marginBottom: spacing.medium,
      }}
    >
      {children}
    </div>
  );
};

const StatefulSwitch = ({ children, ...props }: { children: ReactNode }) => {
  const [checked, setChecked] = useState(false);
  return (
    <Switch checked={checked} onClick={() => setChecked(!checked)} {...props}>
      {children}
    </Switch>
  );
};

export default function FieldsPage() {
  const { spacing } = useTheme();
  return (
    <Page>
      <h1>Form Fields</h1>
      <h2>Text Inputs</h2>
      <FieldWrapper>
        <TextInput placeholder="Small" size="small" />
      </FieldWrapper>
      <FieldWrapper>
        <TextInput placeholder="Medium" size="medium" />
      </FieldWrapper>
      <FieldWrapper>
        <TextInput placeholder="Large" size="large" />
      </FieldWrapper>
      <h3>Disabled</h3>
      <FieldWrapper>
        <TextInput value="This one is disabled" size="medium" disabled />
      </FieldWrapper>
      <h2>TextAreas</h2>
      <FieldWrapper>
        <TextArea />
      </FieldWrapper>
      <h2>Segmented Controls</h2>
      <FieldWrapper>
        <SegmentedControl animate segments={['one', 'two', 'three']} />
      </FieldWrapper>
      <h2>Checkboxes</h2>
      <FieldWrapper>
        <Checkbox css={{ marginRight: spacing.medium }}>Check Me</Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} checked readOnly>
          Always Checked
        </Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} disabled readOnly>
          Disabled Option
        </Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} disabled checked readOnly>
          Checked + Disabled
        </Checkbox>
      </FieldWrapper>
      <h3>Sizes</h3>
      <FieldWrapper>
        <Checkbox css={{ marginRight: spacing.medium }} size="small">
          Small
        </Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} size="medium">
          Medium
        </Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} size="large">
          Large
        </Checkbox>
      </FieldWrapper>
      <h2>Radio Buttons</h2>
      <FieldWrapper>
        <Radio name="radio-set" css={{ marginRight: spacing.medium }}>
          Check Me
        </Radio>
        <Radio name="radio-set" css={{ marginRight: spacing.medium }}>
          Or Me
        </Radio>
        <Radio css={{ marginRight: spacing.medium }} checked readOnly>
          Always Checked
        </Radio>
        <Radio css={{ marginRight: spacing.medium }} disabled readOnly>
          Disabled
        </Radio>
        <Radio css={{ marginRight: spacing.medium }} disabled checked readOnly>
          Checked + Disabled
        </Radio>
      </FieldWrapper>
      <h3>Sizes</h3>
      <FieldWrapper>
        <Radio name="radio-size-set" css={{ marginRight: spacing.medium }} size="small">
          Small
        </Radio>
        <Radio name="radio-size-set" css={{ marginRight: spacing.medium }} size="medium">
          Medium
        </Radio>
        <Radio name="radio-size-set" css={{ marginRight: spacing.medium }} size="large">
          Large
        </Radio>
      </FieldWrapper>
      <h2>Switches</h2>
      <FieldWrapper>
        <StatefulSwitch css={{ marginRight: spacing.medium }}>Switch Me</StatefulSwitch>
        <Switch css={{ marginRight: spacing.medium }} checked>
          Always On
        </Switch>
        <Switch css={{ marginRight: spacing.medium }} checked disabled>
          On + Disabled
        </Switch>
      </FieldWrapper>
    </Page>
  );
}
