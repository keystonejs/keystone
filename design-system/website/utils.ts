export const toLabel = (str: string) => str.substr(0, 1).toUpperCase() + str.substr(1);

const VOWELS = ['a', 'e', 'i', 'o', 'u'];
export const aAn = (before: string) => `a${VOWELS.includes(before.charAt(0)) ? 'n' : ''}`;
