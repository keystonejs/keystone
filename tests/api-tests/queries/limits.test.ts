import { text, integer, relationship } from '@keystone-next/fields';
import { createSchema, list } from '@keystone-next/keystone/schema';
import { setupTestRunner } from '@keystone-next/testing';
import { apiTestConfig, expectGraphQLValidationError, expectLimitsExceededError } from '../utils';
import { depthLimit, definitionLimit, fieldLimit } from './validation';

const runner = setupTestRunner({
  config: apiTestConfig({
    lists: createSchema({
      Post: list({
        fields: {
          title: text(),
          author: relationship({ ref: 'User.posts', many: true }),
        },
      }),
      User: list({
        fields: {
          name: text(),
          favNumber: integer(),
          posts: relationship({ ref: 'Post.author', many: true }),
        },
        graphql: {
          queryLimits: {
            maxResults: 2,
          },
        },
      }),
    }),
    graphql: {
      queryLimits: { maxTotalResults: 6 },
      apolloConfig: {
        validationRules: [depthLimit(3), definitionLimit(3), fieldLimit(8)],
      },
    },
  }),
});

describe('maxResults Limit', () => {
  describe('Basic querying', () => {
    test(
      'users',
      runner(async ({ context }) => {
        const users = await context.lists.User.createMany({
          data: [
            { data: { name: 'Jess', favNumber: 1 } },
            { data: { name: 'Johanna', favNumber: 8 } },
            { data: { name: 'Sam', favNumber: 5 } },
            { data: { name: 'Theo', favNumber: 2 } },
          ],
        });

        // 2 results is okay
        let data = await context.graphql.run({
          query: `
          query {
            allUsers(
              where: { name_contains: "J" },
              orderBy: { name: asc },
            ) {
              name
            }
          }
      `,
        });

        expect(data).toHaveProperty('allUsers');
        expect(data.allUsers).toEqual([{ name: 'Jess' }, { name: 'Johanna' }]);

        // No results is okay
        data = await context.graphql.run({
          query: `
          query {
            allUsers(
              where: { name: "Nope" }
            ) {
              name
            }
          }
      `,
        });

        expect(data).toHaveProperty('allUsers');
        expect(data.allUsers.length).toEqual(0);

        // Count is still correct
        data = await context.graphql.run({
          query: `query { usersCount }`,
        });

        expect(data).toHaveProperty('usersCount');
        expect(data.usersCount).toBe(users.length);

        // This query is only okay because of the "first" parameter
        data = await context.graphql.run({
          query: `
          query {
            allUsers(first: 1) {
              name
            }
          }
      `,
        });

        expect(data).toHaveProperty('allUsers');
        expect(data.allUsers.length).toEqual(1);

        // This query returns too many results
        let errors;
        ({ errors } = await context.graphql.raw({
          query: `
          query {
            allUsers {
              name
            }
          }
      `,
        }));

        expectLimitsExceededError(errors, [{ path: ['allUsers'] }]);

        // The query results don't break the limits, but the "first" parameter does
        ({ errors } = await context.graphql.raw({
          query: `
          query {
            allUsers(
              where: { name: "Nope" },
              first: 100000
            ) {
              name
            }
          }
      `,
        }));

        expectLimitsExceededError(errors, [{ path: ['allUsers'] }]);
      })
    );
  });

  describe('Relationship querying', () => {
    test(
      'posts by user',
      runner(async ({ context }) => {
        const users = await context.lists.User.createMany({
          data: [
            { data: { name: 'Jess', favNumber: 1 } },
            { data: { name: 'Johanna', favNumber: 8 } },
            { data: { name: 'Sam', favNumber: 5 } },
          ],
        });
        await context.lists.Post.createMany({
          data: [
            { data: { author: { connect: [{ id: users[0].id }] }, title: 'One author' } },
            {
              data: {
                author: { connect: [{ id: users[0].id }, { id: users[1].id }] },
                title: 'Two authors',
              },
            },
            {
              data: {
                author: {
                  connect: [{ id: users[0].id }, { id: users[1].id }, { id: users[2].id }],
                },
                title: 'Three authors',
              },
            },
          ],
        });
        // Reset the count for each query
        context.totalResults = 0;
        // A basic query that should work
        let posts = await context.lists.Post.findMany({
          where: { title: 'One author' },
          query: 'title author { name }',
        });

        expect(posts).toEqual([{ title: 'One author', author: [{ name: 'Jess' }] }]);

        // Reset the count for each query
        context.totalResults = 0;
        // Each subquery is within the limit (even though the total isn't)
        posts = await context.lists.Post.findMany({
          where: {
            OR: [{ title: 'One author' }, { title: 'Two authors' }],
          },
          orderBy: { title: 'asc' },
          query: 'title author(orderBy: { name: asc }) { name }',
        });
        expect(posts).toEqual([
          { title: 'One author', author: [{ name: 'Jess' }] },
          { title: 'Two authors', author: [{ name: 'Jess' }, { name: 'Johanna' }] },
        ]);

        // Reset the count for each query
        context.totalResults = 0;
        // This post has too many authors
        let errors;
        ({ errors } = await context.graphql.raw({
          query: `
          query {
            allPosts(
              where: { title: "Three authors" },
            ) {
              title
              author {
                name
              }
            }
          }
      `,
        }));

        expectLimitsExceededError(errors, [{ path: ['allPosts', expect.any(Number), 'author'] }]);

        // Requesting the too-many-authors post is okay as long as the authors aren't returned
        // Reset the count for each query
        context.totalResults = 0;
        posts = await context.lists.Post.findMany({
          where: { title: 'Three authors' },
          query: 'title',
        });

        expect(posts).toEqual([{ title: 'Three authors' }]);

        // Some posts are okay, but one breaks the limit
        // Reset the count for each query
        context.totalResults = 0;
        ({ errors } = await context.graphql.raw({
          query: `
          query {
            allPosts {
              title
              author {
                name
              }
            }
          }
      `,
        }));

        expectLimitsExceededError(errors, [{ path: ['allPosts', expect.any(Number), 'author'] }]);

        // All subqueries are within limits, but the total isn't
        // Reset the count for each query
        context.totalResults = 0;
        ({ errors } = await context.graphql.raw({
          query: `
          query {
            allPosts(where: { title: "Two authors" }) {
              title
              author {
                posts {
                  title
                }
              }
            }
          }
      `,
        }));

        expectLimitsExceededError(errors, [{ path: ['allPosts', 0, 'author', 1, 'posts'] }]);
      })
    );
  });
});

// FIXME: we need upstream support in the graphql package to make KS validation rules work for internal requests
// (Low priority, but makes the API less surprising if rules work everywhere by default.)

describe('maxDepth Limit', () => {
  test(
    'script kiddie',
    runner(async ({ graphQLRequest }) => {
      // Block a script-kiddie attempt to DoS the server with nested queries
      const { body } = await graphQLRequest({
        query: `
            query {
              allPosts {
                author {
                  posts {
                    author {
                      name
                    }
                  }
                }
              }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [{ message: 'Operation has depth 5 (max: 3)' }]);
    })
  );

  test(
    'mutation script kiddie',
    runner(async ({ graphQLRequest }) => {
      const { body } = await graphQLRequest({
        query: `
            mutation {
              updatePost( id: "foo", data: { title: "bar" }) {
                author {
                  posts {
                    author {
                      name
                    }
                  }
                }
              }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [{ message: 'Operation has depth 5 (max: 3)' }]);
    })
  );

  test(
    'one fragment',
    runner(async ({ graphQLRequest }) => {
      // Slightly sneakier depth violation using a fragment
      const { body } = await graphQLRequest({
        query: `
            query nestingbomb {
              allPosts {
                ...f
              }
            }
            fragment f on Post {
              author {
                posts {
                  title
                }
              }
            }
            `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [{ message: 'Operation has depth 4 (max: 3)' }]);
    })
  );

  test(
    'multiple fragments',
    runner(async ({ graphQLRequest }) => {
      // Sneakier violation using multiple fragments
      const { body } = await graphQLRequest({
        query: `
            query nestingbomb {
              allPosts {
                ...f1
              }
            }
            fragment f1 on Post {
              author {
                ...f2
              }
            }
            fragment f2 on User {
              posts {
                title
              }
            }
            `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [{ message: 'Operation has depth 4 (max: 3)' }]);
    })
  );

  test(
    'mutual fragment reference',
    runner(async ({ graphQLRequest }) => {
      // Infinite loop (illegal as GraphQL, but needs to be handled)
      const { body } = await graphQLRequest({
        query: `
            query nestingbomb {
              allPosts {
                ...f1
              }
            }
            fragment f1 on Post {
              author {
                ...f2
              }
            }
            fragment f2 on User {
              posts {
                ...f1
              }
            }
            `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Operation has depth Infinity (max: 3)' },
        { message: 'Operation has depth Infinity (max: 3)' },
        { message: 'Operation has depth Infinity (max: 3)' },
        { message: 'Request contains Infinity fields (max: 8)' },
        { message: 'Cannot spread fragment "f1" within itself via "f2".' },
      ]);
    })
  );

  test(
    'undefined fragment',
    runner(async ({ graphQLRequest }) => {
      // (also illegal as GraphQL, but needs to be handled)
      const { body } = await graphQLRequest({
        query: `
            query {
              allPosts {
                ...nosuchfragment
              }
            }
            `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Undefined fragment "nosuchfragment"' },
        { message: 'Undefined fragment "nosuchfragment"' },
        { message: 'Unknown fragment "nosuchfragment".' },
      ]);
    })
  );
});

describe('maxDefinitions Limit', () => {
  test(
    'queries',
    runner(async ({ graphQLRequest }) => {
      // Too many queries
      const { body } = await graphQLRequest({
        operationName: 'a',
        query: `
            query a {
              allPosts {
                title
              }
            }
            query b {
              allPosts {
                title
              }
            }
            query c {
              allPosts {
                title
              }
            }
            query d {
              allPosts {
                title
              }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Request contains 4 definitions (max: 3)' },
      ]);
    })
  );

  test(
    'fragments',
    runner(async ({ graphQLRequest }) => {
      const { body } = await graphQLRequest({
        operationName: 'q1',
        query: `
            fragment f1 on Post {
              title
            }
            fragment f2 on Post {
              title
            }
            query q1 {
              allPosts {
                ...f1
              }
            }
            query q2 {
              allPosts {
                ...f2
              }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Request contains 4 definitions (max: 3)' },
      ]);
    })
  );

  test(
    'mutations',
    runner(async ({ graphQLRequest }) => {
      const { body } = await graphQLRequest({
        operationName: 'm1',
        query: `
            mutation m1 {
              updatePost(id: "foo", data: { title: "bar" }) {
                title
              }
            }
            mutation m2 {
              updatePost(id: "foo", data: { title: "bar" }) {
                title
              }
            }
            mutation m3 {
              updatePost(id: "foo", data: { title: "bar" }) {
                title
              }
            }
            mutation m4 {
              updatePost(id: "foo", data: { title: "bar" }) {
                title
              }
            }
          `,
      }).expect(400);

      // This isn't the only error, but that's okay
      expectGraphQLValidationError(body.errors, [
        { message: 'Request contains 4 definitions (max: 3)' },
      ]);
    })
  );
});

describe('maxFields Limit', () => {
  test(
    'one operation',
    runner(async ({ graphQLRequest }) => {
      const { body } = await graphQLRequest({
        query: `
            query {
              allPosts {
                title
                author {
                  name
                  favNumber
                }
              }
              allUsers {
                name
                favNumber
                posts {
                  title
                }
              }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Request contains 10 fields (max: 8)' },
      ]);
    })
  );

  test(
    'many operations',
    runner(async ({ graphQLRequest }) => {
      const { body } = await graphQLRequest({
        operationName: 'a',
        query: `
            query a {
              allPosts {
                title
              }
              allUsers {
                name
              }
            }
            query b {
              allPosts {
                title
              }
              allUsers {
                name
              }
            }
            query c {
              allPosts {
                title
              }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Request contains 10 fields (max: 8)' },
      ]);
    })
  );

  test(
    'fragments',
    runner(async ({ graphQLRequest }) => {
      const { body } = await graphQLRequest({
        operationName: 'a',
        query: `
            fragment f on User {
              name
              favNumber
            }
            query a {
              allPosts {
                title
                author {
                  ...f
                }
              }
              users1: allUsers {
                ...f
              }
              users2: allUsers {
                ...f
              }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Request contains 11 fields (max: 8)' },
      ]);
    })
  );

  test(
    'unused fragment',
    runner(async ({ graphQLRequest }) => {
      const { body } = await graphQLRequest({
        operationName: 'a',
        query: `
            fragment unused on User {
              name
              favNumber
            }
            fragment f on User {
              name
              favNumber
            }
            query a {
              allPosts {
                title
                author {
                  ...f
                }
              }
              users1: allUsers {
                ...f
              }
              users2: allUsers {
                ...f
              }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Request contains 13 fields (max: 8)' },
        { message: 'Fragment "unused" is never used.' },
      ]);
    })
  );

  test(
    'billion laughs',
    runner(async ({ graphQLRequest }) => {
      // https://en.wikipedia.org/wiki/Billion_laughs
      const { body } = await graphQLRequest({
        operationName: 'a',
        query: `
            query a {
              u1: allUsers {
                ...lol1
              }
              u2: allUsers {
                ...lol1
              }
              u3: allUsers {
                ...lol1
              }
              u4: allUsers {
                ...lol1
              }
              u5: allUsers {
                ...lol1
              }
            }
            fragment lol1 on User {
              p1: posts {
                ...lol2
              }
              p2: posts {
                ...lol2
              }
              p3: posts {
                ...lol2
              }
              p4: posts {
                ...lol2
              }
              p5: posts {
                ...lol2
              }
            }
            fragment lol2 on Post {
              title
              author { id }
            }
          `,
      }).expect(400);

      expectGraphQLValidationError(body.errors, [
        { message: 'Operation has depth 4 (max: 3)' },
        { message: 'Request contains 105 fields (max: 8)' },
      ]);
    })
  );
});
