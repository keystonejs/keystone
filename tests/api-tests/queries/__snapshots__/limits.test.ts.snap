// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql.maxTake enforces the default 1`] = `
[
  {
    "extensions": {
      "code": "KS_LIMITS_EXCEEDED",
    },
    "locations": [
      {
        "column": 11,
        "line": 3,
      },
    ],
    "message": "Your request exceeded server limits",
    "path": [
      "posts",
    ],
  },
]
`;
